File: android/wearscope/src/main/kotlin/io/wearscope/WSProvisioning.kt
/*
 * Zero-config provisioning — no signup, no dashboard visit before you have data.
 *
 * On first launch without an API key, the SDK asks the server for an anonymous
 * project, stores the credentials locally, and logs the dashboard URL once.
 * The project can later be claimed into an account server-side; nothing is lost.
 */
package io.wearscope

import android.content.Context
import android.util.Log
import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject

internal data class WSCredentials(
    val apiKey: String,
    val projectId: String?,
    val dashboardUrl: String?,
)

internal object WSProvisioning {
  /** Default cloud endpoint — zero-config sends here unless the app overrides it. */
  const val DEFAULT_ENDPOINT = "https://gs.foldalpha.com"

  private const val PREFS = "wearscope"
  private const val KEY_API = "api_key"
  private const val KEY_PROJECT = "project_id"
  private const val KEY_DASHBOARD = "dashboard_url"

  fun cached(context: Context): WSCredentials? {
    val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    val key = prefs.getString(KEY_API, null) ?: return null
    return WSCredentials(key, prefs.getString(KEY_PROJECT, null), prefs.getString(KEY_DASHBOARD, null))
  }

  private fun store(context: Context, creds: WSCredentials) {
    context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
        .putString(KEY_API, creds.apiKey)
        .putString(KEY_PROJECT, creds.projectId)
        .putString(KEY_DASHBOARD, creds.dashboardUrl)
        .apply()
  }

  /**
   * Request an anonymous project. Returns null when the server is unreachable or
   * declines — the caller stays in local mode and retries on the next launch.
   * Must be called off the main thread.
   */
  fun provision(context: Context, endpoint: String, appName: String): WSCredentials? = try {
    val conn = URL("${endpoint.trimEnd('/')}/v1/projects/anonymous").openConnection() as HttpURLConnection
    conn.requestMethod = "POST"
    conn.setRequestProperty("Content-Type", "application/json")
    conn.doOutput = true
    conn.connectTimeout = 15_000
    conn.readTimeout = 15_000
    conn.outputStream.use { it.write(JSONObject().put("name", appName).toString().toByteArray()) }
    val body = if (conn.responseCode in 200..299) conn.inputStream.bufferedReader().readText() else null
    conn.disconnect()
    body?.let {
      val json = JSONObject(it)
      val key = json.optString("api_key")
      if (key.isBlank()) null
      else WSCredentials(
          key,
          json.optString("project_id").ifBlank { null },
          json.optString("dashboard_url").ifBlank { null },
      ).also { creds -> store(context, creds) }
    }
  } catch (_: Exception) {
    null
  }

  /** One-line log banner so a human (or the agent that wired this up) knows where to look. */
  fun announce(creds: WSCredentials, fresh: Boolean) {
    val where = creds.dashboardUrl ?: DEFAULT_ENDPOINT
    Log.i("WearScope", "${if (fresh) "Anonymous project created" else "Reporting"} → $where")
  }
}

File: Sources/WearScope/WSProvisioning.swift
/*
 * Zero-config provisioning — no signup, no dashboard visit before you have data.
 *
 * On first launch without an API key, the SDK asks the server for an anonymous
 * project, stores the credentials locally, and prints the dashboard URL once.
 * The key can later be claimed into an account server-side; nothing is lost.
 */

import Foundation

/// Credentials obtained from (or cached for) anonymous provisioning.
struct WSCredentials: Codable, Sendable {
  let apiKey: String
  let projectId: String?
  let dashboardURL: String?
}

enum WSProvisioning {
  /// Default cloud endpoint — zero-config sends here unless the app overrides it.
  static let defaultEndpoint = URL(string: "https://gs.foldalpha.com")!

  private static var storeURL: URL? {
    guard let dir = try? FileManager.default.url(
      for: .applicationSupportDirectory, in: .userDomainMask,
      appropriateFor: nil, create: true) else { return nil }
    let ours = dir.appendingPathComponent("WearScope", isDirectory: true)
    try? FileManager.default.createDirectory(at: ours, withIntermediateDirectories: true)
    return ours.appendingPathComponent("credentials.json")
  }

  static func cached() -> WSCredentials? {
    guard let url = storeURL, let data = try? Data(contentsOf: url) else { return nil }
    return try? JSONDecoder().decode(WSCredentials.self, from: data)
  }

  private static func store(_ creds: WSCredentials) {
    guard let url = storeURL, let data = try? JSONEncoder().encode(creds) else { return }
    try? data.write(to: url, options: .atomic)
  }

  /// Request an anonymous project. Returns nil when the server is unreachable or
  /// declines — the caller stays in local mode and retries on the next launch.
  static func provision(endpoint: URL, appName: String) async -> WSCredentials? {
    var req = URLRequest(url: endpoint.appendingPathComponent("v1/projects/anonymous"))
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    req.httpBody = try? JSONSerialization.data(withJSONObject: ["name": appName])
    req.timeoutInterval = 15

    guard let (data, resp) = try? await URLSession.shared.data(for: req),
          let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let key = json["api_key"] as? String, !key.isEmpty else { return nil }

    let creds = WSCredentials(
      apiKey: key,
      projectId: json["project_id"] as? String,
      dashboardURL: json["dashboard_url"] as? String)
    store(creds)
    return creds
  }

  /// One-line console banner so a human (or the agent that wired this up) knows where to look.
  static func announce(_ creds: WSCredentials, fresh: Bool) {
    let where_ = creds.dashboardURL ?? defaultEndpoint.absoluteString
    print("[WearScope] \(fresh ? "Anonymous project created" : "Reporting") → \(where_)")
  }
}
